activesupport/lib/active_support/descendants_tracker.rb

```diff
@@ -17,8 +17,21 @@ def direct_descendants(klass)
       end
     end
 
-    if RubyFeatures::CLASS_DESCENDANTS
+    @clear_disabled = false
+
+    if RubyFeatures::CLASS_SUBCLASSES
+      @@excluded_descendants = ObjectSpace::WeakMap.new
```

Comment:
cc @eregon can I rely on `ObjectSpace::WeakMap` having weak keys on Truffle?

It seems like it doesn't (`truffleruby-21.2.0`):
```ruby
>> m = ObjectSpace::WeakMap.new
=> #<ObjectSpace::WeakMap:0x8668>
>> m[Object.new] = 42
=> 42
>> m.size
=> 1
>> 4.times { GC.start }
=> 4
>> m.size
=> 1
>> 4.times { GC.start }
=> 4
>> m.size
=> 1
```

whereas on MRI 3.0.2:
```ruby
>> m = ObjectSpace::WeakMap.new
=> #<ObjectSpace::WeakMap:0x00007feb039d5490>
>> m[Object.new] = 42
=> 42
>> m.size
=> 1
>> 4.times { GC.start }
=> 4
>> m.size
=> 0
```

I know it's not exactly public API, but it's so much more useful that dealing with `WeakRef`...

Reply:
On TruffleRuby `ObjectSpace::WeakMap` only has weak values, not keys.
It's essentially not possible to have both weak keys and weak values on JVM AFAIK, and it's same semantics on JRuby.
`ObjectSpace::WeakMap` is under-specified and it's unclear how it works on CRuby (I once asked and it seems nobody knew), but for portability one needs to consider only `ObjectSpace::WeakMap` values are weak, keys are strong.

If you want/need a map with weak keys, I think it needs to be introduced as a new, separate class.

Reply:
Ok, that's what I though, thanks for confirming. I think this solution is good for MRI, but it could cause leaks in dev mode for Truffle/JRuby. 

I suppose I could just store them in values, I'd just need to find some key that's unique and won't use memory.

Do you know if `object_id` guaranteed to be unique on Truffle/JRuby? Or at least if it's unique enough to be used in this dev scenario?

Reply:
`object_id` is unique on TruffleRuby, it's simply an atomic long incremented for every object which needs an `object_id`.
(it needs to be unique for `ObjectSpace._id2ref`, but _id2ref is really slow and should not be used).

Reply:
Ok, I added a different codepath for Jruby/Truffle. Unfortunately I don't see any decent way to feature test this, so I check `RUBY_ENGINE`.

Reply:
I would suggest to do `if RUBY_ENGINE != "ruby"`, it seems unlikely other Ruby implementations would admit the complexity of having a map of both weak keys and weak values for the GC implementation.

JRuby does have the semantics of weak values only: https://github.com/jruby/jruby/blob/03695f40d1c954cfdcbcef1e433b76e5d16ff068/core/src/main/java/org/jruby/RubyObjectSpace.java#L318

Longer term I think adding `WeakKeysMap` and `WeakValuesMap` (or just use `ObjectSpace::WeakMap`) to Ruby would be best: https://bugs.ruby-lang.org/issues/16038#note-14

Reply:
> I would suggest to do if RUBY_ENGINE != "ruby"

Make sense. I did that and added a bit more doc about what the hell is happening.

> Longer term I think adding WeakKeysMap and WeakValuesMap (or just use ObjectSpace::WeakMap) to Ruby would be best: https://bugs.ruby-lang.org/issues/16038#note-14

Agreed, but that wasn't really the point of that ticket, it only derived there. Originally what I was asking for was equality semantic, but maybe both subjects can be handled in the same ticket.

Either way thanks for the Truffle info!